resource_obmcs_core_security_list.go

```diff
@@ -332,33 +321,59 @@ func (s *SecurityListResourceCrud) buildICMPOptions(conf map[string]interface{})
 }
 
 func (s *SecurityListResourceCrud) buildTCPOptions(conf map[string]interface{}) (opts *baremetal.TCPOptions) {
-	l := conf["tcp_options"].([]interface{})
-	if len(l) > 0 {
-		confOpts := l[0].(map[string]interface{})
+	options := conf["tcp_options"].([]interface{})
+	if len(options) > 0 {
+		sourcePortRange, destinationPortRange := s.buildSourceAndDestinationPortRanges(options)
 		opts = &baremetal.TCPOptions{
-			baremetal.PortRange{
-				Max: uint64(confOpts["max"].(int)),
-				Min: uint64(confOpts["min"].(int)),
-			},
+			destinationPortRange,
+			sourcePortRange,
 		}
 	}
 	return
 }
 
 func (s *SecurityListResourceCrud) buildUDPOptions(conf map[string]interface{}) (opts *baremetal.UDPOptions) {
-	l := conf["udp_options"].([]interface{})
-	if len(l) > 0 {
-		confOpts := l[0].(map[string]interface{})
+	options := conf["udp_options"].([]interface{})
+	if len(options) > 0 {
+		sourcePortRange, destinationPortRange := s.buildSourceAndDestinationPortRanges(options)
 		opts = &baremetal.UDPOptions{
-			baremetal.PortRange{
-				Max: uint64(confOpts["max"].(int)),
-				Min: uint64(confOpts["min"].(int)),
-			},
+			destinationPortRange,
+			sourcePortRange,
+		}
+	}
+	return
+}
+
+func buildPortRange(conf []interface{}) (portRange *baremetal.PortRange) {
+	if len(conf) > 0 && conf[0] != nil {
+		mapConf := conf[0].(map[string]interface{})
+
+		// The ok value will always be true, so we need to check against the default value instead.
+		max := mapConf["max"].(int)
+		min := mapConf["min"].(int)
+
+		// Max and Min default to 0, and that is not a valid port number, so we can assume that if
+		// the value is 0 then the user has not set the port number.
+		if max != 0 || min != 0 {
```

Comment:
Since both max and min are optional, what happens if someone specifies max without min or vice-versa?

Should we assume no min means everything <= max and no max means everything >= min?

Reply:
Specifying one without the other will fail, since the non-specified one will be 0. That's another case that would be nice to check with a validation function, until we do that the service will validate.

Reply:
I guess the question is whether the writer of the tf config knows that not specifying a value means 0. A tf config writer might incorrectly assume that not specifying a value means that it's an open ended range.

Not a blocker, just some food for thought.

Reply:
I'll add a comment about this in the new PR. An important fact that I didn't mention is that 0 is not a valid port number, and I've verified with networking that that will never be a valid input for these params.